lib/Bindings/Python/setup.py

```diff
@@ -0,0 +1,120 @@
+#  Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+#  See https://llvm.org/LICENSE.txt for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Build/install the circt-core python package.
+# Note that this includes a relatively large build of LLVM (~2400 C++ files)
+# and can take a considerable amount of time, especially with defaults.
+# To install:
+#   pip install . --use-feature=in-tree-build
+# To build a wheel:
+#   pip wheel . --use-feature=in-tree-build
+#
+# It is recommended to build with Ninja and ccache. To do so, set environment
+# variables by prefixing to above invocations:
+#   CMAKE_GENERATOR=Ninja \
+#   CMAKE_C_COMPILER_LAUNCHER=ccache \
+#   CMAKE_CXX_COMPILER_LAUNCHER=ccache
+#
+# On CIs, it is often advantageous to re-use/control the CMake build directory.
+# This can be set with the CIRCT_CMAKE_BUILD_DIR env var.
+#
+# By default, this will use the llvm-project submodule included with CIRCT.
+# This can be overridden with the CIRCT_LLVM_DIR env var.
+
+import os
+import shutil
+import subprocess
+
+from distutils.command.build import build as _build
+from setuptools import find_namespace_packages, setup, Extension
+from setuptools.command.build_ext import build_ext
+from setuptools.command.build_py import build_py
+
+
+# Build phase discovery is unreliable. Just tell it what phases to run.
+class CustomBuild(_build):
+
+  def run(self):
+    self.run_command("build_py")
+    self.run_command("build_ext")
+    self.run_command("build_scripts")
+
+
+class CMakeExtension(Extension):
+
+  def __init__(self, name, sourcedir=""):
+    Extension.__init__(self, name, sources=[])
+    self.sourcedir = os.path.abspath(sourcedir)
+
+
+class CMakeBuild(build_py):
+
+  def run(self):
+    target_dir = self.build_lib
+    circt_dir = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "..", "..", ".."))
+    cmake_build_dir = os.getenv("CIRCT_CMAKE_BUILD_DIR")
+    if not cmake_build_dir:
+      cmake_build_dir = os.path.join(circt_dir, "build")
+    cmake_install_dir = os.path.join(cmake_build_dir, "..", "install")
+    llvm_dir = os.getenv("CIRCT_LLVM_DIR")
+    if not llvm_dir:
+      llvm_dir = os.path.join(circt_dir, "llvm")
+    cmake_args = [
+        "-DCMAKE_BUILD_TYPE=Release",  # not used on MSVC, but no harm
+        "-DCMAKE_INSTALL_PREFIX={}".format(os.path.abspath(cmake_install_dir)),
+        "-DLLVM_ENABLE_PROJECTS=mlir",
+        "-DLLVM_EXTERNAL_PROJECTS=circt",
+        "-DLLVM_EXTERNAL_CIRCT_SOURCE_DIR={}".format(circt_dir),
+        "-DLLVM_TARGETS_TO_BUILD=host",
+        "-DMLIR_ENABLE_BINDINGS_PYTHON=ON",
+        "-DCIRCT_BINDINGS_PYTHON_ENABLED=ON",
+    ]
+    build_args = []
```

Comment:
Fyi, the progressively latest hacks I've found to work everywhere: https://github.com/google/iree/blob/1cfe936dbcfeb73984868d955c0a2c372f677fe3/llvm-external-projects/iree-compiler-api/setup.py#L101